docs(recipes): Accelerating Non-Additive Measures (#3297)
cube-js/cube.js: docs/content/Examples-Tutorials-Recipes/Accelerating-Non-Additive-Measures.md

---
title: Accelerating Non-Additive Measures
permalink: /recipes/non-additivity
category: Examples & Tutorials
subCategory: Query acceleration
menuOrder: 1
---

## Use case

We want to run queries against
[pre-aggregations](https://cube.dev/docs/caching#pre-aggregations) only to ensure our
application's superior performance. Usually, accelerating a query is as simple as
including its measures and dimensions to the pre-aggregation
[definition](https://cube.dev/docs/schema/reference/pre-aggregations#parameters-measures).

[Non-additive](https://cube.dev/docs/caching/pre-aggregations/getting-started#ensuring-pre-aggregations-are-targeted-by-queries-non-additivity)
measures (e.g., average values or distinct counts) are a special case.
Pre-aggregations with such measures are less likely to be
[selected](https://cube.dev/docs/caching/pre-aggregations/getting-started#ensuring-pre-aggregations-are-targeted-by-queries-selecting-the-pre-aggregation)
to accelerate a query. However, there are a few ways to work around that.

## Data schema

Let's explore the `Users` cube that contains various measures describing users' age:

> the count of unique age values (`distinctAges`), the average age (`avgAge`), and the
> 90th [percentile](https://cube.dev/docs/recipes/percentiles) of age (`p90Age`).

Reviewer: Maybe put it in bullets to be more easily digestible for readers?

```suggestion

- count of unique age values (`distinctAges`)
- average age (`avgAge`)
- 90th [percentile](https://cube.dev/docs/recipes/percentiles) of age (`p90Age`)

```

```javascript
    distinctAges: {
      sql: `age`,
      type: `countDistinct`,
    },

    avgAge: {
      sql: `age`,
      type: `avg`,
    },

    p90Age: {
      sql: `PERCENTILE_CONT(0.9) WITHIN GROUP (ORDER BY age)`,
      type: `number`,
    },
```

> All of these measures are non-additive. Practically speaking, it means that the following pre-aggregation would accelerate only a query that fully matches its

Reviewer 2: ```suggestion
All of these measures are non-additive. Practically speaking, it means that the following pre-aggregation would only accelerate a query that fully matches its
```

> All of these measures are non-additive. Practically speaking, it means that the following pre-aggregation would only accelerate a query that fully matches its

Reviewer: ```suggestion
All of these measures are non-additive. Practically speaking, it means that the pre-aggregation below would only accelerate a query that fully matches its
```

definition:

```javascript
    main: {
      measures: [
        CUBE.distinctAges,
        CUBE.avgAge,
        CUBE.p90Age
      ],
      dimensions: [
        CUBE.gender
      ]
    },
```

> So, this query will match the pre-aggregation above and, thus, will be accelerated:

Reviewer: ```suggestion
This query will match the pre-aggregation above and, thus, will be accelerated:
```

```javascript
{
  "measures": [
    "Users.distinctAges",
    "Users.avgAge",
    "Users.p90Age"
  ],
  "dimensions": [
    "Users.gender"
  ]
}
```

> Meanwhile, this query won't match the same pre-aggregation because it contains non-addtive measures and omits the `gender` dimension. Thus, it won't be accelerated:

Reviewer: ```suggestion
Meanwhile, the query below won't match the same pre-aggregation because it contains non-additive measures and omits the `gender` dimension. It won't be accelerated:
```

```javascript
{
  "measures": [
    "Users.distinctAges",
    "Users.avgAge",
    "Users.p90Age"
  ]
}
```

> Let's explore possible workarounds.

Reviewer 2: ```suggestion
Let's explore some possible workarounds.
```

### Replacing with approximate additive measures

Often, non-additive `countDistinct` measures can be changed to have the [`countDistinctApprox` type](https://cube.dev/docs/schema/reference/types-and-formats#measures-types-count-distinct-approx)
which will make them additive and orders of magnitude more performant. This
`countDistinctApprox` measures can be used in pre-aggregations. However, there are two
drawbacks:

- This type is approximate, so the measures might yield slightly different results compared to their `countDistinct` counterparts. Please comsult with your database's
documentation to learn more.

> - The `countDistinctApprox` is not supported on all databases. Currently, Cube supports it for Athena, BigQuery, and Snowflake.

Reviewer: ```suggestion
- The `countDistinctApprox` is not supported with all databases. Currently, Cube supports it for Athena, BigQuery, and Snowflake.
```

> - The support for `countDistinctApprox` is dependent on the database. Currently, Cube supports it for BigQuery, Athena, and Snowflake.

Reviewer 2: ```suggestion
- The `countDistinctApprox` is not supported on all databases. Currently, Cube supports it for Athena, BigQuery, and Snowflake.
```

Reviewer 2: We probably want to list the DBs in alphabetical order 😄

For example, the `distinctAges` measure can be rewritten as follows:

```javascript
    distinctAges: {
      sql: `age`,
      type: `countDistinctApprox`,
    },
```

### Decomposing into a formula with additive measures

Non-additive `avg` measures can be rewritten as
[calculated measures](https://cube.dev/docs/schema/reference/measures#calculated-measures)
that reference additive measures only. Then, this additive measures can be used in
pre-aggregations.

For example, the `avgAge` measure can be rewritten as follows:

```javascript
    avgAge: {
      sql: `${CUBE.ageSum} / ${CUBE.count}`,
      type: `number`,
    },

    ageSum: {
      sql: `age`,
      type: `sum`,
    },

    count: {
      type: `count`,
    },
```

### Providing multiple pre-aggregations

> If two workarounds described above don't apply to your use case, feel free to create

Reviewer: ```suggestion
If the two workarounds described above don't apply to your use case, feel free to create
```
